### ISSUE TITLE
AddCaller() Option generates invalid caller

### ISSUE THREAD

**Issue by lukasmalkmus** [uber-go/zap#510]
Hi `zap` maintainers and community. I think I've found an issue, but to be honest, I'm not 100% sure it's related directly to `zap`. It **could** be that I misunderstood related documentation and thus my custom wrapper around `zap` is not implemented correctly.

To get you started, two excerpts from my logs:
```bash
{"level":"info","ts":1507562518.823207,"caller":"auth/main.go:77","msg":"GRPC listening","service":"auth","address":":50051"}
{"level":"info","ts":1507562526.25527,"caller":"<autogenerated>:1","msg":"loading customer","service":"auth","component":"sql","username":"jeff@example.com"}
```

I implemented a small authentication microservice to try out [Jaeger](https://uber.github.io/jaeger/) and [Opentracing](http://opentracing.io). I took some inspiration from the [hotrod](https://github.com/jaegertracing/jaeger/tree/master/examples/hotrod) example and imported its [log implementation](https://github.com/jaegertracing/jaeger/blob/master/examples/hotrod/pkg/log) to add some context to my spans.

I think the issue I have is easy to spot in the log messages presented above: I expected to see the source code line but got `<autogenerated>:1` in the second log message.

Since I use a wrapper around `zap`, I passed the `AddCaller` option to generate my base logger, as suggested in the documentation:
```go
lg, _ := zap.NewProduction(
	zap.AddCallerSkip(1),
	zap.AddStacktrace(zap.FatalLevel),
)
base := log.NewFactory(lg).With(zap.String("service", "auth")) // log package from hotrod example
```

Taking a look at the log output given above, the first log message looks like expected: Instead of seeing something like `log.log:65` (the wrapper) I see the actual caller. But the second log message contains an invalid caller, and I wonder what went wrong.

Could this be related to `zap`?

**Comment by lukasmalkmus** [comment]
Related: https://github.com/golang/go/commit/e97209515ad8c4042f5a3ef32068200366892fc2

**Comment by lylex** [comment]
Same issue. So any workaround for this?

**Comment by akshayjshah** [comment]
There isn't a good workaround for this; it should be fixed in Go 1.10.

Thanks for the detailed report and the follow-up, @lukasmalkmus!

**Comment by akshayjshah** [comment]
Okay, this should be resolved now. Please re-open and let us know if you're still running into this!
